### DISCUSSION TITLE
A convertor option to get the query result as a dictionary when querying a single row or an array of dictionaries when querying multiple rows

### DISCUSSION THREAD

**Discussion by DimitarDevP** [sqlalchemy/sqlalchemy#10149] [Ideas]
I have been working with SQLAlchemy for a few years now and I have been creating my own sets of functions, namely one of the function I have is called `convert` : 

```
def convertor(instance, exclude=[], force_array=False):
    if instance is None:
        return instance
    if type(instance) != list:
        instance = [instance]
        
    ret = []
    for i in instance:
        obj = dict()
        for key in i.__dict__:
            if key not in exclude and key != "_sa_instance_state":
                obj[key] = i.__dict__[key]
        ret.append(obj)
        
    if len(ret) == 1 and force_array == False:
        return ret[0]
    return ret
```

Whenever I want to return the results as an array, I do the following: 
```
languages = Languages.query.all()
languages = convertor(languages, [], True)
```

Which will return all of the languages as an array of dictionaries where the key is the column name and the value of each key is the respective row value for that column:
```
[
 {"id" : 1, "name" : "English"},
 {"id" : 2, "name" : "Macedonian"},
...
]
```

This saves me a lot of time and makes my code a lot cleaner, but I reckon it can be made even simpler by being able to pass an argument in the `all` and `first` methods of the `Query` class to tell it to return the result as a dict or array of dicts:
`languages = Languages.query.all(as_dict=True)`

Furthermore, I have a function that allows for creating hash maps of the table, so there could also be an option to do `hash_by="key"` to tell it to return the result as a hash map. Probably a good idea to add a parametar that would specify open or closed hash. Here is my implementation (this takes into consideration that the result is already a dict or an array of dictionaries):
```
def hash_query_results_obj(array, col_key, cbh=True):
    if type(array) != list:
        array = [array]
    if len(array) == 0:
        return {}
    ret = {}
    for item in array:
        if cbh:
            ret[item[col_key]] = item
        else:
            if ret[item[col_key]] is None:
                ret[item[col_key]] = [item]
            else:    
                ret[item[col_key]].append(item)
    return ret
```

**Comment by CaselIT** [comment]
Hi,

In general these functions are not as simple as you provided. For example your concertor does not handle relationships.

To have a shallow array of dict without relationship, using v2 syntax could be achieved by simply using a connection instead of a session to execute a select and return the mapping result: `session.connection().execute(the_select). mappings().all()`

Overall serialization functionalities are out of scope of sqlalchemy, so I'm -1 on this

**Reply by DimitarDevP** [reply]
Maybe a separate class that would handle these kind of conversions is in order? I am down to write the code myself if the mods are willing to approve my PR.
